Add wharfer exec, go.sum file and simplify README

niklas88 opened the pull request:
`wharfer exec` is added with support for environment variables and `-it` (as with `run` this differs from the separate `-i` and `-t` flags in `docker`). Detaching is also supported.

niklas88 committed:
Add wharfer exec, go.sum file and simplify README

lehmann-4178656ch commented on go.sum:
```diff
@@ -0,0 +1,2 @@
+github.com/docker/docker v1.13.1 h1:5VBhsO6ckUxB0A8CE5LlUJdXzik9cbEbBTQ/ggeml7M=
```
This errors on my machine when trying to build this branch using ```go version go1.11.5 linux/amd64```:
```go: verifying github.com/docker/docker@v1.13.1: checksum mismatch
	downloaded: h1:IkZjBSIc8hBjLpqeAbeE5mca5mNgeatLHBy3GO78BWo=
	go.sum:     h1:5VBhsO6ckUxB0A8CE5LlUJdXzik9cbEbBTQ/ggeml7M=```

lehmann-4178656ch replied:
Also errors on arch linux using ```go version go1.12.1 linux/amd64```
```$ go build
go: finding github.com/docker/docker v1.13.1
go: downloading github.com/docker/docker v1.13.1
verifying github.com/docker/docker@v1.13.1: checksum mismatch
    downloaded: h1:IkZjBSIc8hBjLpqeAbeE5mca5mNgeatLHBy3GO78BWo=
    go.sum:     h1:5VBhsO6ckUxB0A8CE5LlUJdXzik9cbEbBTQ/ggeml7M=```

niklas88 replied:
I forced an update locally and got the same version, this is likely related to this issue
https://github.com/golang/go/issues/29278

niklas88 committed:
Force correct checksum

This is likely caused by this Go bug:
https://github.com/golang/go/issues/29278

lehmann-4178656ch commented on wrap/exec.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	argsLeft := exec.Cmd.NArg()
+	if argsLeft < 2 {
```
argsLeft is only used here, so maybe remove the variable and check the value directly
```suggestion
	if exec.Cmd.Narg() < 2 {
```

lehmann-4178656ch commented on wrap/exec.go:
```diff
@@ -0,0 +1,70 @@
+
+	// The [CONTAINER] positional argument
+	container := exec.Cmd.Args()[0]
+	argsLeft--
```
Variable is not read later
```suggestion
```

lehmann-4178656ch commented on wrap/exec.go:
```diff
@@ -0,0 +1,70 @@
+	// [COMMAND] argument
+	command := exec.Cmd.Args()[1]
+	args = append(args, command)
+	argsLeft--
```
Variable is not read later

lehmann-4178656ch commented on wrap/attach.go:
```diff
@@ -21,7 +21,7 @@ func (c *Attach) ParseToArgs(rawArgs []string) []string {
 		args = append(args, "--")
 		for _, arg := range c.Cmd.Args() {
-			// Only allow attaching to named containers.
+			// If the container is given by name we prepend the user name
```
The new comment does not reflect that `IsHexOnly` arguments are dropped as the append in line 27 is only executed for non-hex identifiers.
This currently ensures that no one can snoop around in containers created by others.

niklas88 replied:
Ok, I think this is acutally wrong in `attach` as we support hex names independent of the username for `kill`.

lehmann-4178656ch commented on main.go:
```diff
@@ -33,6 +33,7 @@ var networkCreate wrap.NetworkCreate
 var networkList wrap.NetworkList
 var networkRemove wrap.NetworkRemove
 var attach wrap.Attach
+var execcmd wrap.Exec
```
Maybe we should discuss the naming and unify all wrapped commands.

lehmann-4178656ch reviewed (commented):
Tried it and it's working, thanks for this feature! Still found some small things which could be changed.

lehmann-4178656ch commented on wrap/exec.go:
```diff
@@ -0,0 +1,70 @@
+	// The [CONTAINER] positional argument
+	container := exec.Cmd.Args()[0]
+	argsLeft--
+	if !IsHexOnly(container) {
```
Maybe allow only named containers, otherwise one could exec commands in containers created by others. See `attach` command and my comment there.

niklas88 replied:
I think the plan was to allow hex names even if they are for other users

niklas88 committed:
Address review, allow attach with hex name
